### PR TITLE
#13 Tracing bullet: added R2DBC with tests and example domain

### LINKED ISSUE #13: Add a database to the project
- Docker compose with postgres should be added before
- Add some library for migrations
- JOOQ to use database from spring application
- JOOQ classes should be generated automatically
- Write some dummy code at layers api-service-repository to test and verify project structure
- Use [Test Containers](https://testcontainers.com) for testing

### PR DESCRIPTION
**Pull request by vityaman**

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -1,2 +1,3 @@
 [*.{kt,kts}]
 ktlint_code_style = intellij_idea
+ktlint_standard_no-wildcard-imports = disabled
```
Comment (vityaman) [inline .editorconfig:3]: Detekt detects it, but more easier to configure. Also we can consider to get rid of ktlint completely. Will create a ticket to backlog.

File: .github/workflows/gradle.yml [hunk 2]
```diff
@@ -22,7 +22,3 @@ jobs:
 
       - name: Build Gateway
         run: (cd gateway && ./gradlew :build)
-
-      - name: Test Gateway
-        run: (cd gateway && ./gradlew :koverHtmlReport)
```
Comment (vityaman) [inline .github/workflows/gradle.yml:27]: Tests are ran at `gradle :build `, so we don't need it.

File: config/detekt.yml [hunk 3]
```diff
@@ -68,6 +68,10 @@ exceptions:
   ObjectExtendsThrowable:
     active: true
 
+formatting:
+  NoWildcardImports:
+    active: false
```
Comment (vityaman) [inline config/detekt.yml:73]: Duplicates `style/WildcardImport`

File: gateway/docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+    image: postgres
+    restart: always
+    ports:
+      - "5432:5432"
+    volumes:
+      - ./src/main/resources/database:/docker-entrypoint-initdb.d
+    environment:
+      POSTGRES_DB: postgres
+      POSTGRES_USER: postgres
+      POSTGRES_PASSWORD: postgres
```
Comment (vityaman) [inline gateway/docker-compose.yml:14]: A database for a local playground.

File: gateway/src/main/kotlin/ru/itmo/lms/gateway/GatewayApplication.kt [hunk 5]
```diff
@@ -1,9 +1,14 @@
 package ru.itmo.lms.gateway
 
 import org.springframework.boot.autoconfigure.SpringBootApplication
+import org.springframework.boot.autoconfigure.jooq.JooqAutoConfiguration
 import org.springframework.boot.runApplication
 
-@SpringBootApplication
+@SpringBootApplication(
+    exclude = [
+        JooqAutoConfiguration::class,
```
Comment (vityaman) [inline gateway/src/main/kotlin/ru/itmo/lms/gateway/GatewayApplication.kt:9]: Without this it is not working :) As I understand we provide this configuration from `r2dbc` connection poll.

File: gateway/src/main/kotlin/ru/itmo/lms/gateway/domain/model/Note.kt [hunk 6]
```diff
@@ -0,0 +1,11 @@
+package ru.itmo.lms.gateway.domain.model
+
+data class Note(
+    val id: Id,
+    val content: String,
+) {
+    @JvmInline
+    value class Id(val value: Long)
```
Comment (vityaman) [inline gateway/src/main/kotlin/ru/itmo/lms/gateway/domain/model/Note.kt:8]: Newtype pattern.

File: gateway/src/main/kotlin/ru/itmo/lms/gateway/storage/jooq/JooqNoteStorage.kt [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
+    override suspend fun create(note: Note.Draft): Note {
+        val sql = dsl
+            .insertInto(NOTE, NOTE.CONTENT)
+            .values(note.content)
+            .returningResult(NOTE.ID, NOTE.CONTENT)
+
+        return Mono.from(sql)
+            .map { row -> NoteRecord(row.value1()!!, row.value2()!!) }
```
Comment (vityaman) [inline gateway/src/main/kotlin/ru/itmo/lms/gateway/storage/jooq/JooqNoteStorage.kt]: Just don't know how to do it clear.

File: gateway/src/main/kotlin/ru/itmo/lms/gateway/storage/jooq/entity/NoteMapping.kt [hunk 8]
```diff
@@ -0,0 +1,15 @@
+package ru.itmo.lms.gateway.storage.jooq.entity
+
+import ru.itmo.lms.gateway.domain.model.Note
+import ru.itmo.lms.gateway.storage.jooq.tables.records.NoteRecord
+
+fun NoteRecord.toModel(): Note =
+    Note(
+        id = Note.Id(this.id!!),
```
Comment (vityaman) [inline gateway/src/main/kotlin/ru/itmo/lms/gateway/storage/jooq/entity/NoteMapping.kt:8]: It is called after retrieval from the database, so all fields must be set.

File: gateway/src/main/kotlin/ru/itmo/lms/gateway/storage/jooq/JooqNoteStorage.kt (hunk 7 above)
Comment (k1b24) [reply gateway/src/main/kotlin/ru/itmo/lms/gateway/storage/jooq/JooqNoteStorage.kt]: Better practice is to retrieve values by attributes names row.get("column_name", Class::class.java)

Maybe this function will be helpful here
https://github.com/k1b24/itmo-blps/blob/main/kachalka-backend/src/main/kotlin/ru/itmo/util/daoUtils.kt#L6-L7C1

On the other hand you use code gen so maybe it won't be so clear to appeal to columns by it's names

**Pull request merged**